src/librustc_typeck/outlives/implicit_infer.rs

```diff
@@ -0,0 +1,203 @@
+) {
+    /*
+        // inferred_outlives_predicate = ['b: 'a] // after round 2
+        struct Foo<'a, 'b> {
+            bar: Bar<'a, 'b> // []
+            // round 2: ['b: 'a] in `required_predicates`
+        }
+
+        // inferred_outlives_predicate = ['b: 'a] // updated after round 1
+        struct Bar<'a, 'b> {
+            field &'a &'b u32 // ['b: 'a] // added to `required_predicates`
+        } // required_predicates = ['b: 'a]
+    */
+
+    let mut changed = true;
+    while changed {
+        changed = false;
+        let mut visitor = InferVisitor {
+            tcx: tcx,
+            inferred_outlives_map: &mut inferred_outlives_map,
+            changed: changed,
+        };
+
+        //iterate over all the crates
+        tcx.hir.krate().visit_all_item_likes(&mut visitor);
+    }
+}
+
+pub struct InferVisitor<'cx, 'tcx: 'cx> {
+    tcx: TyCtxt<'cx, 'tcx, 'tcx>,
+    inferred_outlives_map: &'cx mut FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>>,
+    changed: bool,
+}
+
+impl<'cx, 'tcx> ItemLikeVisitor<'tcx> for InferVisitor<'cx, 'tcx> {
+    fn visit_item(&mut self, item: &hir::Item) {
+        let def_id = self.tcx.hir.local_def_id(item.id);
+
+        let node_id = self.tcx
+            .hir
+            .as_local_node_id(def_id)
+            .expect("expected local def-id");
+        let item = match self.tcx.hir.get(node_id) {
+            hir::map::NodeItem(item) => item,
+            _ => bug!(),
+        };
+
+        let mut local_required_predicates = FxHashMap();
+        match item.node {
+            hir::ItemUnion(..) |
+            hir::ItemEnum(..) |
+            hir::ItemStruct(..) => {
+                let adt_def = self.tcx.adt_def(def_id);
+                for field_def in adt_def.all_fields() {
+                    let required_predicates = required_predicates_to_be_wf(self.tcx, field_def);
+                    local_required_predicates.extend(required_predicates);
+                }
+            }
+            _ => {}
+        };
+
+        if local_required_predicates.len() > 0 {
+            self.changed = true;
+            self.inferred_outlives_map.extend(local_required_predicates);
+        }
+    }
+
+    fn visit_trait_item(&mut self, trait_item: &'tcx hir::TraitItem) {}
+
+    fn visit_impl_item(&mut self, impl_item: &'tcx hir::ImplItem) {}
+}
+
+//FIXME This is custom calculation that to figure out what predicates need to be added
+fn required_predicates_to_be_wf<'tcx>(
+    tcx: TyCtxt<'_, 'tcx, 'tcx>,
+    field_def: &ty::FieldDef,
+) -> FxHashMap<DefId, Rc<Vec<ty::Predicate<'tcx>>>> {
+
+    let mut predicates = FxHashMap();
+    // Get the type of the field with identity substs applied.  For
+    // now, let's just see if that causes horrible cycles.
+    let ty = tcx.type_of(field_def.did);
+    match ty.sty {
+        // For each type `&'a T`, we require `T: 'a`
+        ty::TyRef(r, mt) => {
+            let outlives_pred = ty::Binder(ty::OutlivesPredicate(mt.ty, r))
```

Comment:
@nikomatsakis can you verify that `mt.ty` is the correct Ty to pass here and not `ty`. There is not much documentation for [OutlivesPredicates](https://github.com/rust-lang/rust/blob/90eb44a5897c39e3dff9c7e48e3973671dcd9496/src/librustc/ty/mod.rs#L1105)

Reply:
I can verify that, yes. If you have `T = &'a U`, then `mt.ty` here represents `U` and `ty` represents `T` (and `r` represents `'a`). So we want `U: 'a`.